provisioning/baremetal_secrets.go

```diff
@@ -0,0 +1,126 @@
+package provisioning
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"math/big"
+
+	"golang.org/x/crypto/bcrypt"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	coreclientv1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+const (
+	baremetalSecretName = "metal3-mariadb-password"
+	baremetalSecretKey  = "password"
+	ironicUsernameKey   = "username"
+	ironicPasswordKey   = "password"
+	ironicHtpasswdKey   = "htpasswd"
+	ironicConfigKey     = "auth-config"
+	ironicSecretName    = "metal3-ironic-password"
+	ironicUsername      = "ironic-user"
+	inspectorSecretName = "metal3-ironic-inspector-password"
+	inspectorUsername   = "inspector-user"
+)
+
+func generateRandomPassword() (string, error) {
+	chars := []rune("ABCDEFGHIJKLMNOPQRSTUVWXYZ" +
+		"abcdefghijklmnopqrstuvwxyz" +
+		"0123456789")
+	length := 16
+	buf := make([]rune, length)
+	numChars := big.NewInt(int64(len(chars)))
+	for i := range buf {
+		c, err := rand.Int(rand.Reader, numChars)
+		if err != nil {
+			return "", err
+		}
+		buf[i] = chars[c.Uint64()]
+	}
+	return string(buf), nil
+}
+
+// CreateMariadbPasswordSecret creates a Secret for Mariadb password
+func CreateMariadbPasswordSecret(client coreclientv1.SecretsGetter, targetNamespace string) error {
+	_, err := client.Secrets(targetNamespace).Get(context.Background(), baremetalSecretName, metav1.GetOptions{})
+	if !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	// Secret does not already exist. So, create one.
+	password, err := generateRandomPassword()
+	if err != nil {
+		return err
+	}
+	_, err = client.Secrets(targetNamespace).Create(
+		context.Background(),
+		&corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      baremetalSecretName,
+				Namespace: targetNamespace,
+			},
+			StringData: map[string]string{
+				baremetalSecretKey: password,
+			},
+		},
+		metav1.CreateOptions{},
+	)
+	return err
+}
+
+// CreateIronicPasswordSecret creates a Secret for the Ironic Password
+func CreateIronicPasswordSecret(client coreclientv1.SecretsGetter, targetNamespace string) error {
+	return createIronicSecret(client, targetNamespace, ironicSecretName, ironicUsername, "ironic")
+}
+
+// CreateInspectorPasswordSecret creates a Secret for the Ironic Inspector Password
+func CreateInspectorPasswordSecret(client coreclientv1.SecretsGetter, targetNamespace string) error {
+	return createIronicSecret(client, targetNamespace, inspectorSecretName, inspectorUsername, "inspector")
+}
+
+func createIronicSecret(client coreclientv1.SecretsGetter, targetNamespace string, name string, username string, configSection string) error {
+	_, err := client.Secrets(targetNamespace).Get(context.Background(), name, metav1.GetOptions{})
```

Comment:
Should we be using the pattern that the MAO uses, of having utility functions to create resources and then generic functions to ensure those resources exist?

https://github.com/openshift/machine-api-operator/blob/master/pkg/operator/sync.go#L320

Reply:
Maybe that wouldn't apply in the case of the secrets, since we don't want to change them every time?

Reply:
Maybe we could. I don't understand everything that is happening in  https://github.com/openshift/library-go/blob/master/pkg/operator/resource/resourceapply/core.go#L237 tbh.

Reply:
Couple of things to consider:
1. Do we want to make all 3 Secrets immutable?
https://github.com/kubernetes/api/blob/master/core/v1/types.go#L5624
(I don't understand the scenario where we would want to just change a secret and want metal3 or any other to be re-deployed)
2. I could use https://github.com/openshift/library-go/blob/master/pkg/operator/resource/resourceapply/core.go#L237 to actually create the Secrets. The pattern pointed to earlier (https://github.com/openshift/cluster-baremetal-operator/pull/29#discussion_r498977454) is not possible exactly because the concept of generations doesn't seem to exist for Secret types.

Reply:
How will making the secrets immutable affect upgrades from clusters that do not have immutable secrets?

Using `ApplySecret()` looks like it's probably a good idea.

Reply:
Maybe we can create a separate ticket to track the idea of making those secrets immutable, to give us time to do the research without blocking this work?

Reply:
I don't see any reason not to make them immutable (provided that feature gate is actually enabled in OpenShift). It won't affect existing secrets but at least new ones will be. Note that the Secret can still be deleted, at which point we'll create a new one, so it doesn't change a lot.

`ApplySecret()` seems to be about updating secrets, so not what we want.

Reply:
Yes, ApplySecrets() does not work for us. (Its name is misleading because as @zaneb mentioned, it is about updating secrets.)
This is why:
1. We 1st have to create a Secret type
2. Call ApplySecret() with this Secret
3. ApplySecret would get an existing Secret with the same name as the one passed in. If found, it updates it with the new value.
4. If not found, it applies the Secret passed in.

Step 3 is problematic for the way we want our secrets to be manged. So, sticking with the current implementation.